sample-app/src/components/experiment/experiment.tsx

```diff
@@ -62,15 +65,53 @@ const LegacyExperiment = () => {
     },
     dispatch: globalDispatch,
   } = useGlobal()
+  const { setMessages } = useMessageController()
 
-  const validationViolations: ValidationViolations = useMemo(
+  const violations: ValidationViolations = useMemo(
     () => validateExperiment(experiment),
     [experiment]
   )
 
+  useEffect(() => {
```

Comment:
Haven't found a way to avoid this `useEffect`. Removing it results in a [Cannot update a component while rendering a different component warning](https://stackoverflow.com/questions/62336340/cannot-update-a-component-while-rendering-a-different-component-warning)

Reply:
The way to remove it (generally) is to trigger the `setMessages` invocation from an event handler somewhere, e.g., call a central `doValidation` function on all events that potentially cause a change. - for now we can keep this useEffect